### DISCUSSION TITLE
RFC: Adopting mimalloc as the Allocator in Wezterm

### DISCUSSION THREAD

**Discussion by Uzaaft** [wezterm/wezterm#4137] [Ideas]
# RFC: Adopting mimalloc as the Allocator in Wezterm

## Introduction

This RFC proposes that `wezterm`, a GPU-accelerated terminal emulator, change its current memory allocator to `mimalloc`, a general-purpose allocator with excellent performance characteristics.

Benefits of `mimalloc`:

1. **Performance**: `mimalloc` consistently outperforms many other allocators in benchmarks. The reduced fragmentation and efficient allocation strategies can result in a noticeable performance gain.
  
2. **Concurrency**: It has first-class support for concurrent programs, which can be highly beneficial for multi-threaded operations in `wezterm`.

3. **Safety**: `mimalloc` takes precautions against heap corruption and other vulnerabilities, making it a safer choice. We could make this an opt-in feature behind a flag.

4. **Compactness**: It tends to produce less memory overhead and fragmentation, which is always a boon for memory-intensive applications.

## Technical Details

The integration of `mimalloc` would involve:

1. Adding the `mimalloc` library as a dependency.
2. Changing the allocator setup in the build and runtime system of `wezterm`.

Potential roadblocks:

1. **Integration Challenges**: There might be parts of `wezterm` that make assumptions about the current allocator, and these would need to be identified and addressed.

2. **Compatibility**: While `mimalloc` is designed to be a drop-in replacement, there could be platform-specific issues that need to be resolved.

## Impact on Current Users

For current users of `wezterm`, this change should be mostly transparent. They would stand to benefit from any performance or memory usage improvements, but they should not need to change anything about their current setup or usage of the software.

## Backward Compatibility

Given that changing the allocator shouldn't affect the external APIs or functionalities, backward compatibility shouldn't be a concern. However, rigorous testing would ensure that there are no unforeseen regressions.

## Alternatives Considered

Other modern allocators such as `jemalloc` or `tcmalloc` could also be considered. However, `mimalloc` has shown significant promise in recent benchmarks, making it an attractive option.

## Conclusion

Adopting `mimalloc` as the allocator for `wezterm` can potentially enhance its performance and memory efficiency. Given the importance of these metrics for terminal emulators, it's a change worth exploring. Rigorous testing would be vital to ensure that the change is beneficial and does not introduce new issues.

**Comment by wez** [comment]
FWIW, a quick look at the mimalloc rust repo issues shows lots of soundness related issues with no resolution, and there are several forks of differing levels of quality. I am not enthusiastic about integrating something that seems to be so immature and that doesn't have stable ownership/leadership in the rust community.

I'm open to the idea of using `jemalloc` instead, via `jemallocator`, which has a strong and proven track record.

**Reply by Uzaaft** [reply]
I’m open to that as well! Didnt know about the rust soundness issues, I’ve mostly been using autogenerated bindings to the C/C++ API.
I’ll look into integrating jemalloc, should we do it behind a opt in flag?

**Reply by wez** [reply]
Thanks. I vote no to an opt-in-flag to start; I'd rather know about any issues very quickly and use that to decide whether it is worth the headache of maintaining an alternative build configuration.  I'd suggest setting up the allocator in the env-bootstrap crate.

**Reply by Uzaaft** [reply]
Great.
Before I start:
Do you have a benchmark suit that could be suitable to measure a baseline, and see if we get improvements or regressions in the behaviour of the program?

**Reply by wez** [reply]
There's no automated setup for that.  I would suggest looking at the issues in the issue tracker that talk about memory usage and see if you can reduce those down to a couple of repeatable scenarios with reliable metrics.  Measurement of memory usage is always challenging, and the characteristics are different between Windows, Linux, the *BSDs, and macOS.

**Comment by Uzaaft** [comment]
While implementing jemalloc into wezterm, I saw this:
https://rust-lang.github.io/rfcs/1183-swap-out-jemalloc.html#:~:text=Using%20jemalloc%20implies%20that%20Rust,method%20to%20configure%20the%20allocator.

Would this also be the case for when wezterm itself? I.e the host(wezterm) is using jemalloc, and the child processes are using other allocators?
